examples/bluetooth/ble_remote_control/main/controller.c

```diff
@@ -0,0 +1,381 @@
+{
+    uint8_t button_pin_num = (uint8_t) arg;
+    uint64_t current_time = esp_timer_get_time(); 
+    
+    // set bit whenever a button is pressed
+    // do not send to queue yet in case of other button presses or multiple presses
+    set_button_bit(button_pin_num, 1);
+
+    // Simple button debouncing
+    if (current_time - prev_trigger_time > DEBOUNCE_TIME_US) {
+        input_event_t button_event = {
+            .input_source = INPUT_SOURCE_BUTTON,
+            .data_button = button_input,
+        };
+
+        prev_trigger_time = current_time;
+        xQueueSendFromISR(input_queue, &button_event, 0);
+        button_input = 0; 
+    }
+}
+
+/**
+ * @brief   Set the bit of the button that is pressed
+ * 
+ * @param[in]   pin_num     Button pin number
+ * @param[in]   pin_level   Button pin level (0 or 1)
+*/
+void set_button_bit(int pin_num, int pin_level)
+{
+    switch (pin_num) {
+    case PIN_BUTTON_BOOT: // fallthrough
+    case PIN_BUTTON_A:
+        button_input |= (pin_level << 0);
+        break;
+    case PIN_BUTTON_B:
+        button_input |= (pin_level << 1);
+        break;
+    case PIN_BUTTON_C:
+        button_input |= (pin_level << 2);
+        break;
+    case PIN_BUTTON_D:  
+        button_input |= (pin_level << 3);
+        break;
+    default:    
+        break;
+    }
+}
+
+esp_err_t button_init(void)
+{   
+    const gpio_config_t pin_config = {
+        .pin_bit_mask = BUTTON_PIN_BIT_MASK,
+        .mode = GPIO_MODE_INPUT,
+        .pull_up_en = GPIO_PULLUP_ENABLE,
+        .pull_down_en = GPIO_PULLDOWN_DISABLE,
+        .intr_type = GPIO_INTR_POSEDGE
+    };
+
+    ESP_ERROR_CHECK(gpio_config(&pin_config));
+
+    ESP_ERROR_CHECK(gpio_install_isr_service(ESP_INTR_FLAG_LEVEL1));
+
+#ifdef CONFIG_BUTTON_INPUT_MODE_BOOT
+    ESP_ERROR_CHECK(gpio_isr_handler_add(PIN_BUTTON_BOOT, button_isr_handler, (void*) PIN_BUTTON_BOOT));
+#else // CONFIG_BUTTON_INPUT_MODE_GPIO
+    ESP_ERROR_CHECK(gpio_isr_handler_add(PIN_BUTTON_A, button_isr_handler, (void*) PIN_BUTTON_A));
+    ESP_ERROR_CHECK(gpio_isr_handler_add(PIN_BUTTON_B, button_isr_handler, (void*) PIN_BUTTON_B));
+    ESP_ERROR_CHECK(gpio_isr_handler_add(PIN_BUTTON_C, button_isr_handler, (void*) PIN_BUTTON_C));
+    ESP_ERROR_CHECK(gpio_isr_handler_add(PIN_BUTTON_D, button_isr_handler, (void*) PIN_BUTTON_D));
+#endif
+
+    esp_timer_early_init();
+    return ESP_OK;
+}
+
+void button_read(uint8_t *button_in) 
+{
+#ifdef CONFIG_BUTTON_INPUT_MODE_BOOT
+    // Boot button is active low (high when released, low when pressed)
+    set_button_bit(PIN_BUTTON_BOOT, 1 - gpio_get_level(PIN_BUTTON_BOOT));
+#else // CONFIG_BUTTON_INPUT_MODE_GPIO
+    set_button_bit(PIN_BUTTON_A, gpio_get_level(PIN_BUTTON_A));
+    set_button_bit(PIN_BUTTON_B, gpio_get_level(PIN_BUTTON_B));
+    set_button_bit(PIN_BUTTON_C, gpio_get_level(PIN_BUTTON_C));
+    set_button_bit(PIN_BUTTON_D, gpio_get_level(PIN_BUTTON_D));
+#endif
+
+    *button_in = button_input;
```

Comment:
There is no operation on button_input in this function, yet it is being assigned to *button_in. This will look strange. You may have the current value of button_input returned by set_button_bit. In fact, the button_input can be encapsulated by get/set/clear functions. There is no need for the caller to manage it.